// gmail-node-mailer/dist/config/gmailServiceAccountConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gmailServiceAccountConfig = void 0;
/**
 * Manages configuration for the Gmail service account.
 */
class GmailServiceAccountConfig {
    constructor() {
        this._serviceAccount = undefined;
        this._serviceAccountPath = undefined;
    }
    /**
     * Retrieves the currently set Gmail service account.
     * @returns {IGmailServiceAccount | undefined} The current service account or undefined if not set.
     */
    getServiceAccount() {
        return this._serviceAccount;
    }
    /**
     * Sets the Gmail service account.
     * @param {IGmailServiceAccount | undefined} account The service account to set.
     */
    setServiceAccount(account) {
        this._serviceAccount = account;
    }
    /**
     * Retrieves the path to the Gmail service account file.
     * @returns {string | undefined} The current service account file path or undefined if not set.
     */
    getServiceAccountPath() {
        return this._serviceAccountPath;
    }
    /**
     * Sets the path to the Gmail service account file.
     * @param {string | undefined} path The file path to set.
     */
    setServiceAccountPath(path) {
        this._serviceAccountPath = path;
    }
}
exports.gmailServiceAccountConfig = new GmailServiceAccountConfig();
